pypeit/coadd3d.py

```diff
@@ -0,0 +1,1772 @@
+            if self.blaze_spline is None:
+                self.blaze_wave, self.blaze_spec = wave_spl, spec_spl
+                self.blaze_spline = interp1d(wave_spl, spec_spl, kind='linear',
+                                             bounds_error=False, fill_value="extrapolate")
+
+    def set_voxel_sampling(self):
+        """
+        This function checks if the spatial and spectral scales of all frames are consistent.
+        If the user has not specified either the spatial or spectral scales, they will be set here.
+        """
+        # Make sure all frames have consistent pixel scales
+        ratio = (self._spatscale[:, 0] - self._spatscale[0, 0]) / self._spatscale[0, 0]
+        if np.any(np.abs(ratio) > 1E-4):
+            msgs.warn("The pixel scales of all input frames are not the same!")
+            spatstr = ", ".join(["{0:.6f}".format(ss) for ss in self._spatscale[:,0]*3600.0])
+            msgs.info("Pixel scales of all input frames:" + msgs.newline() + spatstr + "arcseconds")
+        # Make sure all frames have consistent slicer scales
+        ratio = (self._spatscale[:, 1] - self._spatscale[0, 1]) / self._spatscale[0, 1]
+        if np.any(np.abs(ratio) > 1E-4):
+            msgs.warn("The slicer scales of all input frames are not the same!")
+            spatstr = ", ".join(["{0:.6f}".format(ss) for ss in self._spatscale[:,1]*3600.0])
+            msgs.info("Slicer scales of all input frames:" + msgs.newline() + spatstr + "arcseconds")
+        # Make sure all frames have consistent wavelength sampling
+        ratio = (self._specscale - self._specscale[0]) / self._specscale[0]
+        if np.any(np.abs(ratio) > 1E-2):
+            msgs.warn("The wavelength samplings of the input frames are not the same!")
+            specstr = ", ".join(["{0:.6f}".format(ss) for ss in self._specscale])
+            msgs.info("Wavelength samplings of all input frames:" + msgs.newline() + specstr)
+
+        # If the user has not specified the spatial scale, then set it appropriately now to the largest spatial scale
+        if self._dspat is None:
+            self._dspat = np.max(self._spatscale)
+            msgs.info("Adopting a square pixel spatial scale of {0:f} arcsec".format(3600.0 * self._dspat))
+        # If the user has not specified the spectral sampling, then set it now to the largest value
+        if self._dwv is None:
+            self._dwv = np.max(self._specscale)
+            msgs.info("Adopting a wavelength sampling of {0:f} Angstrom".format(self._dwv))
+
+    def load(self):
+        """
+        This is the main function that loads in the data, and performs several frame-specific corrections.
+        If the user does not wish to align or combine the individual datacubes, then this routine will also
+        produce a spec3d file, which is a DataCube representation of a PypeIt spec2d frame for SlicerIFU data.
+        """
+        # Initialise variables
+        wave_ref = None
```

Comment:
Loading from files should be a class method. But these frame specific transformations should probably be in a core method.

Reply:
Okay as a compromise, please add the following to the documentation to make the control flow more clear. Add a list of all of the class attributes that are assigned by this function, i.e. all the self.X assignments. I agree that return values would be combersome and dilly, but it is needs to be clear what attributes this method assigns withoiut having to read through 500 lines of code.